statsmodels/tsa/vecm/vecm.py

```diff
@@ -0,0 +1,2146 @@
+        (dimensions assuming no deterministic terms are given)
+    delta_y_1_T : ndarray (neqs x nobs)
+        :math:`(y_1, \\ldots, y_T) - (y_0, \\ldots, y_{T-1})`
+    y_min1 : ndarray (neqs x nobs)
+        (dimensions assuming no deterministic terms are given)
+        :math:`(y_0, \\ldots, y_{T-1}`
+
+    Returns
+    -------
+    result : tuple
+        A tuple of two ndarrays. (See p. 292 in [1]_ for the definition of
+        R_0 and R_1.)
+
+    References
+    ----------
+    .. [1] Lutkepohl, H. 2005. *New Introduction to Multiple Time Series Analysis*. Springer.
+    """
+
+    # todo: rewrite m such that a big (TxT) matrix is avoided
+    m = np.identity(T) - (
+        delta_x.T.dot(inv(delta_x.dot(delta_x.T))).dot(delta_x))  # p. 291
+    r0 = delta_y_1_T.dot(m)  # p. 292
+    r1 = y_min1.dot(m)
+    return r0, r1
+
+
+def _sij(delta_x, delta_y_1_T, y_min1):
+    """Returns matrices and eigenvalues and -vectors used for parameter
+    estimation and the calculation of a models loglikelihood.
+
+    Parameters
+    ----------
+    delta_x : ndarray (diff_lags*neqs x nobs)
+        (dimensions assuming no deterministic terms are given)
+    delta_y_1_T : ndarray (neqs x nobs)
+        :math:`(y_1, \\ldots, y_T) - (y_0, \\ldots, y_{T-1})`
+    y_min1 : ndarray (neqs x nobs)
+        (dimensions assuming no deterministic terms are given)
+        :math:`(y_0, \\ldots, y_{T-1}`
+
+    Returns
+    -------
+    result : tuple
+        A tuple of five ndarrays as well as eigenvalues and -vectors of a
+        certain (matrix) product of some of the returned ndarrays.
+        (See pp. 294-295 in [1]_ for more information on
+        :math:`S_0, S_1, \\lambda_i, \\v_i` for
+        :math:`i \\in \\{1, \\dots, K\\}`.)
+
+    References
+    ----------
+    .. [1] Lutkepohl, H. 2005. *New Introduction to Multiple Time Series Analysis*. Springer.
+    """
+    T = y_min1.shape[1]
+    r0, r1 = _r_matrices(T, delta_x, delta_y_1_T, y_min1)
+    s00 = np.dot(r0, r0.T) / T
+    s01 = np.dot(r0, r1.T) / T
+    s10 = s01.T
+    s11 = np.dot(r1, r1.T) / T
+    s11_ = inv(_mat_sqrt(s11))
+    # p. 295:
+    s01_s11_ = np.dot(s01, s11_)
+    eig = np.linalg.eig(chain_dot(s01_s11_.T, inv(s00), s01_s11_))
+    lambd = eig[0]
+    v = eig[1]
+    # reorder eig_vals to make them decreasing (and order eig_vecs accordingly)
+    lambd_order = np.argsort(lambd)[::-1]
+    lambd = lambd[lambd_order]
+    v = v[:, lambd_order]
+    return s00, s01, s10, s11, s11_, lambd, v
+
+
+def select_coint_rank(endog_tot, det_order, k_ar, method="trace", signif=0.95,
+                      verbose=True):
+    """
+    Calculate the cointegration rank of a VECM.
+
+    Parameters
+    ----------
+    endog_tot : array-like (nobs_tot x neqs)
+        The data with presample.
+    det_order : int
+        * -1 - no deterministic terms
+        * 0 - constant term
+        * 1 - linear trend
+        * >1 - higher polynomial order
```

Comment:
I don't see the context right now.
I think the convenience function for coint rank should raise an exception for all cases where we cannot determine the rank because we are missing the critical values.

The cointegration test should return nan for critical values and warn but not raise. In that case it can still be useful for the user to get the test statistic if the user has access to other tables, or wants to try bootstrap. (I think I have seen bootstrap articles, but didn't read any.)
Also the plan would be to add other approximation for the cointegration test critical/p values at some point in the future. (e.g. I looked at some other articles last summer like one or two by Doornik)